src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -227,25 +228,97 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
-
             List<Tuple<string, string>> queryParametersList = new List<Tuple<string, string>>(sharedQueryParametersList);
             if (progress.ContinuationToken != null)
             {
                 queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.ContinuationToken, progress.ContinuationToken));
             }
 
-            if (_exportJobRecord.ExportType == ExportJobType.Patient)
+            var requestedResourceTypes = _exportJobRecord.ResourceType?.Split(',');
+            var filteredResources = new HashSet<string>();
+            if (_exportJobRecord.Filters != null)
             {
-                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, KnownResourceTypes.Patient));
+                foreach (var filter in _exportJobRecord.Filters)
+                {
+                    filteredResources.Add(filter.ResourceType.ToUpper(CultureInfo.InvariantCulture));
+                }
             }
-            else if (_exportJobRecord.ExportType == ExportJobType.All && !string.IsNullOrEmpty(_exportJobRecord.ResourceType))
+
+            IAnonymizer anonymizer = IsAnonymizedExportJob() ? await CreateAnonymizerAsync(cancellationToken) : null;
+
+            if (!progress.FilteredSearchesComplete)
             {
-                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, _exportJobRecord.ResourceType));
+                var index = -1;
+                if (_exportJobRecord.Filters != null)
+                {
+                    foreach (var filter in _exportJobRecord.Filters)
+                    {
+                        index++;
+
+                        if (progress.CurrentFilter != null && filter != progress.CurrentFilter)
+                        {
+                            continue;
+                        }
+
+                        if (requestedResourceTypes != null &&
+                            requestedResourceTypes.Contains(filter.ResourceType, StringComparison.OrdinalIgnoreCase) &&
+                            (_exportJobRecord.ExportType == ExportJobType.All || filter.ResourceType.Equals(KnownResourceTypes.Patient, StringComparison.Ordinal)))
+                        {
+                            progress.SetFilter(filter);
+
+                            List<Tuple<string, string>> filterQueryParametersList = new List<Tuple<string, string>>(queryParametersList);
+                            foreach (var param in filter.Parameters)
+                            {
+                                filterQueryParametersList.Add(param);
+                            }
+
+                            await SearchWithFilter(exportJobConfiguration, progress, filter.ResourceType, filterQueryParametersList, sharedQueryParametersList, anonymizer, "filter" + index + "-", cancellationToken);
+
+                            progress.SetFilter(null);
+                            await UpdateJobRecordAsync(cancellationToken);
```

Comment:
If the worker crashes and we have to restart export from this state, `progress.CurrentFilter` will be `null`. This means we will end up re-processing all the filters that we had already completed. We should maybe have a different way of tracking completed filters. Could be a different list or remove from existing list?

Reply:
Yea, this method just keeps having holes. I'll make a list of completed filters since each progress record needs its own list.